### PR TITLE
[ENG-229] - Change CAISO Load file URL

### PR DESCRIPTION
**Pull request by Kladar**
## Summary 
This changes the approach to CAISO load file getting slightly. Pulls from the `latest` file first, given that it could be more up to date than the history file if today's data is desired. Also adds type annotations to the function and variables, switches to using f-strings, some other python best practice stuff.

## Details
@kmax12 Mostly gauging how much latitude there is to alter the functionality of this code (for example, removing the `try - except` block, with the desire in the future to handle exceptions separately, given that there are some other failure modes I could see this running into)

## Tests
Tried running just the `tests/test_caiso.py` and it seems to have hung a bit, as well as running into a certificate error that might be straightforwardly remedied. Details:

CleanShot 2024-09-23 at 15 35 36@2x

### CODE REVIEW

**Comment by Kladar** [comment]
Also a good chance for me to see the checks in this repo and build some intuition there, so I'll watch them and ask any questions I might have.

File: gridstatus/caiso.py [hunk 1]
```diff
@@ -1460,19 +1460,19 @@ def _make_timestamp(time_str, today, timezone="US/Pacific"):
...
-            msg = f"Fetching URL: {url}"
-            log(msg, verbose)
-
+def _get_historical(
+    file: str,
+    date: str | pd.Timestamp,
+    verbose: bool = False,
+) -> pd.DataFrame:
+
+    if date == "today":
```
Comment (kmax12) [inline gridstatus/caiso.py:1469]: it's possible `date` doesn't come through as the string today. I think you'll want to us `utils.is_today(date, CAISO.default_timezone)`

**Comment by kmax12** [comment]
> @kmax12 Mostly gauging how much latitude there is to alter the functionality of this code (for example, removing the try - except block, with the desire in the future to handle exceptions separately, given that there are some other failure modes I could see this running into)

you have latitude to do what you see as best. 

in this case, I think it's perfectly fine to remove the try except

File: gridstatus/caiso.py [hunk 2]
```diff
@@ -1460,19 +1460,19 @@ def _make_timestamp(time_str, today, timezone="US/Pacific"):
...
-        log(msg, verbose)
-    except Exception:
-        # fallback if today and no historical file yet
-        if utils.is_today(date, CAISO.default_timezone):
-            url = _BASE + "/%s.csv" % file
-            msg = f"Fetching URL: {url}"
-            log(msg, verbose)
-
+def _get_historical(
+    file: str,
```
Comment (kmax12) [inline gridstatus/caiso.py:1464]: thanks for adding typing. i'd love to do it across more of the codebase as we have time!

**Review by kmax12 — COMMENTED** [review]
this change is what I expected 👌 . one comment on how to check for today

the local test failures seems like a local configuration issue. this is what I get running your branch locally

image

File: gridstatus/caiso.py (hunk 2 above)
Comment (Kladar) [reply gridstatus/caiso.py:1464]: Sounds good - for now I'll make a habit to add it to functions as I touch them.

File: gridstatus/caiso.py (hunk 1 above)
Comment (Kladar) [reply gridstatus/caiso.py]: Done

File: gridstatus/caiso.py [hunk 3]
```diff
@@ -1460,19 +1460,19 @@ def _make_timestamp(time_str, today, timezone="US/Pacific"):
...
+    date: str | pd.Timestamp,
+    verbose: bool = False,
+) -> pd.DataFrame:
+
+    if utils.is_today(date, CAISO.default_timezone):
```
Comment (WillKoehrsen) [inline gridstatus/caiso.py:1469]: What happens if the file for today is not available yet?

Comment (Kladar) [reply gridstatus/caiso.py:1469]: Conceivably there is always a file available at the https://www.caiso.com/outlook/sp/demand.csv URL, which we assume is for "today" (not sure how valid that assumption is, tho)

Comment (WillKoehrsen) [reply gridstatus/caiso.py:1469]: Ahh, okay, I see, thanks. It's the historical file for a given date that may not be available yet.

**Comment by Kladar** [comment]
Thought maybe these build unit tests were failing for some reason of adding types to the utils or something, but looks like they are also failing on `main` for me. 
CleanShot 2024-09-23 at 17 20 58@2x

**Comment by Kladar** [comment]
> this change is what I expected 👌 . one comment on how to check for today
> 
> the local test failures seems like a local configuration issue. this is what I get running your branch locally
> image

Yep, this was a new computer certificate issue, resolved now. 
CleanShot 2024-09-24 at 13 04 23@2x

**Comment by Kladar** [comment]
Woo woo, all CAISO tests passing locally 
CleanShot 2024-09-24 at 13 29 26@2x

**Comment by WillKoehrsen** [comment]
All the CAISO tests and utils tests are passing locally for me as well. Did you figure out why the utils tests failed for you? Those test failures shouldn't be related to the certificate issue since they aren't making any requests.

**Comment by Kladar** [comment]
> All the CAISO tests and utils tests are passing locally for me as well. Did you figure out why the utils tests failed for you? Those test failures shouldn't be related to the certificate issue since they aren't making any requests.

`test_utils` and `test_base` also working for me. I did have to install Java for the last couple CAISO tests (sneaking the JVM in there for some PDF parsing I see haha) 

CleanShot 2024-09-25 at 08 37 12@2x

CleanShot 2024-09-25 at 08 38 46@2x

**Comment by Kladar** [comment]
Looks like the build tests that are failing are 35 for ERCOT and 1 for MISO (meaning the change in this PR seems to be good). Going to transition to running gs-etl with the branch preview as a deployment test.

**Review by kmax12 — APPROVED** [review]

**Pull request merged**